source/common/http/http1/codec_impl.cc

```diff
@@ -407,34 +413,54 @@ void RequestEncoderImpl::encodeHeaders(const RequestHeaderMap& headers, bool end
   encodeHeadersBase(headers, absl::nullopt, end_stream);
 }
 
+int ConnectionImpl::setAndCheckCallbackStatus(Status&& status) {
+  codec_status_ = std::move(status);
+  return codec_status_.ok() ? enumToInt(HttpParserCode::Success) : enumToInt(HttpParserCode::Error);
+}
+
+int ConnectionImpl::setAndCheckCallbackStatusOr(Envoy::StatusOr<int>&& statusor) {
+  if (statusor.ok()) {
+    return statusor.value();
+  } else {
+    codec_status_ = std::move(statusor.status());
+    return enumToInt(HttpParserCode::Error);
+  }
+}
+
 http_parser_settings ConnectionImpl::settings_{
     [](http_parser* parser) -> int {
-      static_cast<ConnectionImpl*>(parser->data)->onMessageBeginBase();
-      return 0;
+      auto status = static_cast<ConnectionImpl*>(parser->data)->onMessageBeginBase();
+      return static_cast<ConnectionImpl*>(parser->data)
+          ->setAndCheckCallbackStatus(std::move(status));
```

Comment:
nit: use local variable to hold static_cast<ConnectionImpl*>(parser->data) for readability.

auto* conn_impl = static_cast<ConnectionImpl*>(parser->data);
auto status = conn_impl->onMessageBeginBase();
return conn_impl->setAndCheckCallbackStatus(std::move(status));

or

auto* conn_impl = static_cast<ConnectionImpl*>(parser->data);
return conn_impl->setAndCheckCallbackStatus(conn_impl->onMessageBeginBase());

Reply:
like that a lot better. @yanavlasov for H/2